### ISSUE TITLE
UpdateConditionNotSatisfied exception and stopped retries when using AzureStorage TimeoutPersister

### ISSUE THREAD

**Issue by particularbot** [Particular/NServiceBus.Persistence.AzureStorage#22]
**Issue by [kseniya-terekhova](https://github.com/kseniya-terekhova)** _1/24/2016 1:28:44 PM +00:00_ _Originally opened as https://github.com/Particular/NServiceBus/issues/3352_

---

We've encountered an issue in our environment when running several instances of NserviceBus with timeouts persisted in AzureStorage. Sometimes message's retries are prematurely stoped after such an exception have been logged:  
System.Data.Services.Client.DataServiceClientException: {"odata.error":{"code":"UpdateConditionNotSatisfied","message":{"lang":"en-US","value":"The update condition specified in the request was not satisfied...

This situation seems to be caused by concurrently working NServiceBus.Azure.TimeoutPersister's copies, that perform following actions:
1. Instance1 calls TryRemove method that loads TimeoutDataEntity with PartionKey==timeoutId, just before Instance2 have removed this entity. Method goes further and conceivably slows down when downloading timeout state from AzureBlob.
2. At that time Instance2 removes all TimeoutDataEntities, performs unsuccessful message handling and schedules next retry by saving TimeoutDataEntity with PartitionKey=='Time in PartitionKeyScope format' and RowKey==timeoutId, and also new TimeoutDataEntity with PartionKey==timeoutId
3. TryRemove method of Instance1 loads "fresh" TimeoutDataEntity with with PartitionKey=='Time in PartitionKeyScope format' (due to insufficient format granularity ("yyyyMMddHH")) and marks it for deletion in ServiceContext. Old TimeoutDataEntity with PartionKey==timeoutId is market for deletion only after that.
4. Instance1 calls non-transactional context.SaveChangesWithRetries(), that removes "fresh" TimeoutDataEntity with PartitionKey=='Time in PartitionKeyScope format', but fails to remove old, non-existent TimeoutDataEntity with PartionKey==timeoutId, throwing UpdateConditionNotSatisfied exception.
5.  TimeoutDataEntity with with PartitionKey=='Time in PartitionKeyScope format' is not found by GetNextChunk method any more, so message retries are stopped. In addition, timeout state from AzureBlob is lost as well.

We found this problem in NserviceBus 4.7.0.0, as in later versions TryRemove method silently swallows DataServiceRequestException and no indication of retries' problem is visible.

The suggestion is to move deletion of TimeoutDataEntity with PartionKey==timeoutId just after it has been loaded and remove AzureBlob with timeout state only when all TimeoutDataEntities were successfully erased.

**Comment by particularbot** [comment]
**Comment by [danielmarbach](https://github.com/danielmarbach)** _https://github.com/Particular/NServiceBus/issues/3352#issuecomment-174465080_

----

@Particular/azure-maintainers can someone please chime in and maybe move this issue over to the corresponding repository?

**Comment by Scooletz** [comment]
@yvesgoeleven Did you take a look at this? Any results?

**Comment by yvesgoeleven** [comment]
This is indeed still an issue, to properly solve it we'd require master election so that only 1 instance is actively managing the data (and the others are in stand by)

Master election would require a core change https://github.com/Particular/NServiceBus/issues/2602, or timeouts would need to be decoupled from the core https://github.com/Particular/NServiceBus.AzureStorageQueues/issues/166

but as of today this issue is still blocked

**Comment by yvesgoeleven** [comment]
@Scooletz Can we close this once we have native delay delivery?

**Comment by Scooletz** [comment]
Yes @yvesgoeleven

**Comment by yvesgoeleven** [comment]
Closing in favor of https://github.com/Particular/PlatformDevelopment/issues/1026
